include/cuco/detail/hyperloglog/hyperloglog.cuh

```diff
@@ -0,0 +1,340 @@
+   *
+   * @note This function synchronizes the given stream.
+   *
+   * @param stream CUDA stream this operation is executed in
+   *
+   * @return Approximate distinct items count
+   */
+  [[nodiscard]] std::size_t estimate(cuco::cuda_stream_ref stream) const
+  {
+    // TODO remove test code
+    // std::size_t* result;
+    // cudaMallocHost(&result, sizeof(std::size_t));
+
+    // int grid_size  = 0;
+    // int block_size = 0;
+    // // TODO check cuda error?
+    // cudaOccupancyMaxPotentialBlockSize(
+    //   &grid_size, &block_size, &cuco::hyperloglog_ns::detail::estimate<ref_type<>>);
+
+    // cuco::hyperloglog_ns::detail::estimate<<<grid_size, block_size, 0, stream>>>(
+    //   result, this->ref());
+    // stream.synchronize();
+
+    // return *result;
+
+    // TODO this function currently copies the registers to the host and then finalizes the result;
+    // move computation to device? Edit: host computation is faster -.-
+    storage_type registers;
+    // TODO check if storage is host accessible
+    CUCO_CUDA_TRY(cudaMemcpyAsync(
+      &registers, this->storage_.get(), sizeof(storage_type), cudaMemcpyDeviceToHost, stream));
+    stream.synchronize();
+
+    using fp_type = typename ref_type<>::fp_type;
+    fp_type sum   = 0;
+    int zeroes    = 0;
+
+    // geometric mean computation + count registers with 0s
+    for (int i = 0; i < registers.size(); ++i) {
+      auto const reg = registers[i];
+      sum += fp_type{1} / static_cast<fp_type>(1ull << reg);
+      zeroes += reg == 0;
+    }
+
+    // pass intermediate result to finalizer for bias correction, etc.
+    return cuco::hyperloglog_ns::detail::finalizer<Precision>::finalize(sum, zeroes);
+  }
+
+  /**
+   * @brief Get device ref.
+   *
+   * @return Device ref object of the current `distinct_count_estimator` host object
+   */
+  [[nodiscard]] ref_type<> ref() const noexcept
+  {
+    return ref_type<>{*(this->storage_.get()), this->hash_};
+  }
+
+  /**
+   * @brief Get storage ref.
+   *
+   * @return Reference to storage
+   */
+  [[nodiscard]] storage_type& storage_ref() const noexcept { return *(this->storage_.get()); }
+
+  /**
+   * @brief Get hash function.
+   *
+   * @return The hash function
+   */
+  [[nodiscard]] auto hash() const noexcept { return this->hash_; }
+
+ private:
+  struct storage_deleter {
+    using pointer = typename storage_allocator_type::value_type*;
+
+    storage_deleter(storage_allocator_type& a) : allocator{a} {}
+
+    storage_deleter(storage_deleter const&) = default;
+
+    void operator()(pointer ptr) { allocator.deallocate(ptr, 1); }
+
+    storage_allocator_type& allocator;
+  };
+
+  Hash hash_;                                               ///< Hash function used to hash items
+  storage_allocator_type storage_allocator_;                ///< Storage allocator
+  storage_deleter storage_deleter_;                         ///< Storage deleter
+  std::unique_ptr<storage_type, storage_deleter> storage_;  ///< Storage
```

Comment:
```suggestion
  storage_type storage_;  ///< Storage
```
All those details should be wrapped in the storage type

Reply:
Not sure if there is a benefit in exposing the `register_type` to the user. If we switch to a packed storage, the "register" would consist of 6 bits in a word. We can't express that as a type. Since the storage has a static size, I think allocating it as `__shared__ ref_type::storage_type storage;` is the most intuitive way.